File: shakespeare/Dna.py
import random
import string


def random_char():
    c = random.randint(32, 126)  # Adjusted range to match JavaScript's 32-127
    return chr(c)


class DNA:

    def __init__(self, length):
        self.genes = [];
        self.fitness = 0;
        for i in range(length):
            self.genes.append(random_char())
    def get_phrase(self):
        return "".join(self.genes)
    def calculate_fitness(self, target):
        score = 0
        for i in range(len(target)):
            if target[i] == self.genes[i]:
                score += 1
        self.fitness = score / len(target)

    def crossover(self, partner):
        child = DNA(len(self.genes))
        midpoint = random.randint(0, len(self.genes))
        child.genes[:midpoint] = self.genes[:midpoint]
        child.genes[midpoint:] = partner.genes[midpoint:]
        return child

    def mutate(self, rate):
        for i in range(len(self.genes)):
            if random.random() < rate:
                self.genes[i] = random_char()

File: shakespeare/Dna.js
export function random(lower, upper = 0) {
  if (upper == 0) {
    return Math.random() * lower;
  }
  return Math.random() * (upper - lower) + lower;
}
export function floor(n) {
  return Math.floor(n);
}
function randomCharacter() {
  let c = floor(random(32, 127));
  return String.fromCharCode(c);
}
let x = [];
for (let i = 0; i < 10; i++) {
  x[i] = randomCharacter()
}

export class DNA {
  constructor(length) {
    this.genes = [];
    this.fitness = 0;
    for (let i = 0; i < length; i++) {
      this.genes[i] = randomCharacter();
    }
  }
  getPhrase() {
    return this.genes.join("");
  }
  getFitness() {
    return this.fitness;
  }
  calculateFitness(target) {
    let score = 0;
    for (let i = 0; i < this.genes.length; i++) {
      if (this.genes[i] == target.charAt(i)) {
        score++;
      }
    }
    this.fitness = score / target.length;
  }

  crossover(partner) {
    let child = new DNA(this.genes.length);
    let midpoint = floor(random(this.genes.length));

    for (let i = 0; i < this.genes.length; i++) {
      if (i < midpoint) {
        child.genes[i] = this.genes[i];

      } else {
        child.genes[i] = partner.genes[i];
      }
    }
    return child;
  }
  mutate(mutationRate) {
    for (let i = 0; i < this.genes.length; i++) {
      if (random(1) < mutationRate) {
        this.genes[i] = randomCharacter();
      }
    }
  }
}
